cmake/CompilerWarnings.cmake

```diff
@@ -4,7 +4,7 @@
 
 # Helper function to enable compiler warnings for a specific set of files
 function(set_file_warnings)
-    option(WARNINGS_AS_ERRORS "Treat compiler warnings as errors" TRUE)
```

Comment:
Is these a reason the default has to be changed?

Reply:
Users compiling SFML do not necessarily want warnings to be treated as errors if it's a third-party library.

If that's intended, they're free to use a special case `-DWARNINGS_AS_ERRORS=1`.

This was not a problem before because gcc did not report warnings, and clang did but they were accidentally ignored. When I left this option as TRUE and allowed clang users to throw errors on warnings, all clang-based CI failed. Mostly because of macOS deprecation warnings and those similar.

I think WARNINGS_AS_ERRORS=1 could be an option for CI (after clang warnings are removed). As far as a user is concerned, the library should default to not include `-Werror` as it is most useful to SFML developers.

Reply:
My understanding was that `WARNINGS_AS_ERRORS` only applies to SFML itself, and that seems true from my testing. It shouldn't propagate to user projects including SFML. But it should be enabled by default for SFML itself.

Reply:
@reconndev When we write code, we can never foresee any platform/architecture specific problems that can arise when building for certain platforms that we didn't have access to when writing the code. The compiler that is building the code knows best if some code might be problematic on a specific architecture hence the warnings that it emits. This can be seen when we didn't take alignment on ARM architectures into account when first building with warnings.

Ideally when building SFML no warnings should ever be emitted regardless what architecture it is built on. Because compiler warnings are very conservative and any false positives are treated as serious bugs by the compiler maintainers one can be fairly confident that they are basically always correct.

The real issue here is that Apple has (as usual) chosen to misuse something that was originally meant as an aid for C++ developers. If a feature is deprecated from the C++ standard (like `std::auto_ptr`) to be removed in the future, that is a case where a warning from a _C++ compiler_ would be OK. Because this happens over the course of many years, and the necessary adjustments to the code are fairly obvious, they are warnings that everybody happily accepts. Apple has a tradition of deprecating APIs fairly frequently and expects developers to change perfectly working code to deal with the warnings. This is a business decision and has nothing to do with ensuring the code does what it is supposed to do (as with the alignment warning on ARM).

Imagine if a user were to really build SFML as a dependency on some fairly new architecture. Warnings would be emitted informing the user that something doesn't feel right. It turns out that some code does invoke undefined behaviour on said platform. The user files a bug report with us. Conventionally we don't ask them if they noticed if any warnings were generated when they were building SFML as a dependency (I'm sure less strict libraries generate a heap of other warnings as well). This would be a really annoying bug to fix, since most of us will not have access to this platform to do our own testing.

In order to reduce the maintenance burden on us (having to ask if the library was "correctly" built) whenever an issue is opened, we have the expectation that the library is built without errors/warnings. If the user chooses to ignore warnings, they do so at their own risk and we can't really support them under those circumstances.

If building on macOS is causing warnings to be generated, then that has to be fixed as opposed to changing the default for all other platforms as well. Whether it involves disabling the deprecation warnings or if all else fails then just changing the default when building for macOS. We can't let business policies impact what developers on all other platforms have to deal with as well.

Reply:
> My understanding was that WARNINGS_AS_ERRORS only applies to SFML itself, and that seems true from my testing.

Did you test it after my fix on a clang compiler? Otherwise, you wouldn't have seen the difference because this simply did not work.

In short, if I (as a user) wanted to enable `WARNINGS_AS_ERRORS` it wouldn't have worked for clang. Now, I fixed that but since Clang has lots of warnings it fails on CI.

Reply:
@binary1248 I don't think you know what this PR is about.

I fixed `WARNINGS_AS_ERRORS` for clang. SFML build ignored warnings on clang before anyway because the flag didn't work properly. Do you expect me to fix all ~200 warnings on macOS in a PR that only wanted to fix `WARNINGS_AS_ERRORS`?

I changed the default value to FALSE so that the CI doesn't fail on all clang-based platforms because *only now* it errors out on warnings and there are many unfixed warnings.

Reply:
Seems like there has been a bit of a misunderstanding 😄 

We can't / don't really want to disable the default behavior, otherwise all new changes, won't be properly checked, thus can easily introduce new warnings, that we need to provide additional fixes again.

There isn't really an easy way to get around fixing all the remaining errors for clang, if you want to utilize the CI to indicate whether the PR actually fixed the issues.

We of course don't expect you to fix all the errors, thus my question whether you're willing to help with the fixes, just so we know, whether we should pick it up and nobody is doing redundant work. 😅 

As for macOS, it might be that we're not able to fix all the warnings, which I think is what binary1248 was pointing to, because Apple marked a lot of APIs as deprecated, which we still want to make use of. So we'll have to find ways to silence those or maybe some can actually be updated to newer APIs.

Reply:
Bugs are instances where actual behaviour doesn't correspond intended behaviour. Yes, it is correct that the actual "bug" is that the ordering of the flags was wrong, and that would be what needs to be fixed. Changing a default for all platforms isn't part of the fix for the actual "bug".

If fixing the issue causes CI to fail for macOS then maybe those warnings would have to be dealt with in a separate PR before this one is merged. If the warnings would have lead to build failure in the initial PR by @vittorioromeo where the sweeping changes were made to resolve warnings on all other platforms the PR wouldn't have been merged. It was an oversight, but not something that makes it necessary to revert the default back now.

I'm not a fan of workarounds for problems that can be fixed with a proportionate amount of effort. Going from memory, many of the deprecation warnings were specific to OpenGL/OpenAL. Those warnings can be disabled on a file by file basis, and there aren't as many distinct files as the warnings might have one believe.

Reply:
This file list is from the CI log of the last failure due to deprecation warnings:
```
examples/cocoa/CocoaAppDelegate.mm
src/SFML/Audio/ALCheck.cpp
src/SFML/Audio/AudioDevice.cpp
src/SFML/Audio/Sound.cpp
src/SFML/Audio/SoundBuffer.cpp
src/SFML/Audio/SoundRecorder.cpp
src/SFML/Audio/SoundSource.cpp
src/SFML/Audio/SoundStream.cpp
src/SFML/Window/OSX/SFApplication.m
src/SFML/Window/OSX/SFApplicationDelegate.m
src/SFML/Window/OSX/SFContext.mm
src/SFML/Window/OSX/SFKeyboardModifiersHelper.mm
src/SFML/Window/OSX/SFOpenGLView+mouse.mm
src/SFML/Window/OSX/SFOpenGLView.h
src/SFML/Window/OSX/SFOpenGLView.mm
src/SFML/Window/OSX/SFViewController.mm
src/SFML/Window/OSX/SFWindowController.mm
src/SFML/Window/OSX/WindowImplCocoa.hpp
src/SFML/Window/OSX/WindowImplDelegateProtocol.h
src/SFML/Window/OSX/cg_sf_conversion.mm
```

It seems the easiest way to disable warnings in these files by adding the following lines somewhere near the top of each:
```cpp
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```

Reply:
Since I did not know about the "no warnings" policy you have at SFML (especially considering I was getting so many warnings before), I wasn't sure whether the default was correct, too. That's why I changed it.

My understanding was that most projects allow some warnings to happen and so CI instances could make use of `-DWARNINGS_AS_ERRORS=1` on e.g. GCC builds, which were clean to not introduce any new ones.

I didn't want to fix all the warnings in one PR. I thought it would be easier gradually. Because of the "no warnings" policy, I will have to do it in this PR or not at all to pass CI checks so thank you for the clarification now, though lengthily delivered.

Reply:
This might also tie in to #1896, the current idea is that there will be a header that is used only internally when building SFML. These `#pragma`s could be placed in there and apply only when building on macOS.

Reply:
> Because of the "no warnings" policy, I will have to do it in this PR or not at all to pass CI checks

I agree that "extending PRs" like this is not really in our interest, it just increases the time-to-merge of the first half of the feature, leading to subsequent merge conflicts and needless extra work.

What if we disabled the treat-warnings-as-errors for now, but immediately open a new issue/PR where we revert it and fix the MacOS errors? That way, we can take our time, and we don't block the existing parts of this PR from becoming part of `master`.

Reply:
Changed the default to FALSE only when CMake declares `APPLE`.

Now, I am working on resolving clang-based compiler errors. The last part will be with deprecations from Apple but I see @binary1248 already did some work on that so I only expect to merge that branch or cherry-pick commits.